clients/pkg/promtail/promtail.go

```diff
@@ -139,14 +148,44 @@ func (p *Promtail) reloadConfig(cfg *config.Config) error {
 			return err
 		}
 		cfg.PositionsConfig.ReadOnly = true
+	} else if cfg.WAL.Enabled {
+		// TODO: Once we refactor all client instantiations into the manager, all ifs statements here will be removed
+		// and p.client will contain the manager, with all instantiated clients inside.
+		p.client, err = client.NewManager(
+			p.metrics,
+			p.logger,
+			cfg.LimitsConfig.MaxStreams,
+			cfg.LimitsConfig.MaxLineSize.Val(),
+			cfg.LimitsConfig.MaxLineSizeTruncate,
+			p.reg,
+			cfg.WAL,
+			cfg.ClientConfigs...,
+		)
+		if err != nil {
+			return err
+		}
 	} else {
 		p.client, err = client.NewMulti(p.metrics, p.logger, cfg.LimitsConfig.MaxStreams, cfg.LimitsConfig.MaxLineSize.Val(), cfg.LimitsConfig.MaxLineSizeTruncate, cfg.ClientConfigs...)
 		if err != nil {
 			return err
 		}
 	}
 
-	tms, err := targets.NewTargetManagers(p, p.reg, p.logger, cfg.PositionsConfig, p.client, cfg.ScrapeConfig, &cfg.TargetConfig)
+	// these are the sinks were all scraped log entries should get to
+	var entryHandlers = []api.EntryHandler{p.client}
+
+	// If WAL is enabled, instantiate the WAL itself, it's writer, and use that as entries sink for all scraping targets
+	if cfg.WAL.Enabled {
+		p.walWriter, err = wal.NewWriter(cfg.WAL, p.logger, p.reg)
+		if err != nil {
+			return fmt.Errorf("failed to create wal writer: %w", err)
+		}
+		entryHandlers = append(entryHandlers, p.walWriter)
+	}
+
+	p.entriesFanout = utils.NewFanoutEntryHandler(timeoutUntilFanoutHardStop, entryHandlers...)
+
+	tms, err := targets.NewTargetManagers(p, p.reg, p.logger, cfg.PositionsConfig, p.entriesFanout, cfg.ScrapeConfig, &cfg.TargetConfig)
```

Comment:
last comment, I think; it feels to me like the creation of the WAL writer should be above within the previous `else if cfg.WAL.Enabled`, and we could reassign the fanout writer to `p.client`

if the usage of the actual Client type for `p.client` makes that difficult lets just find a way to combine these two if statements so that we don't make our job harder when refactoring, we could make `p.walWriter` a pointer, and check if the pointer isn't null when deciding whether to create the entries fanout?

Reply:
Yhea also wasn't fond of that double if now that you mention. Moved a bit the code so that we have a single if branch related to wal-enabled behaviour. I think we can keep the fanout if wal is disabled, since it adds the graceful shutdown which is not there otherwise.

Also, we should keep a Client implementing interface in `p.client`, since it's part of the interface the main promtail objects exposes. In the future, this will be just the `client.Manager`.

Let me know wdyt!